Standalone GNUmakefile and supporting changes

dmcnaugh opened the pull request:
As per #53 

With the following notes:
### ***THIS INTRODUCES A BREAKING CHANGE IN THE EXITING BUILD SYSTEM***

> However this change is easy to accomodate with the addition of `-I.` into the `gcc` command string (see changes to `z80sim/srcsim/Makefile.osx` and `imsaisim/srcssim/Makefile.osx` for examples).

This PR introduces a **new** build system based on a **single** `Makefile` for each simulation (otherwise called a **machine**) that supports all O/S (otherwise called a **platform**. Actually only BSD, CygWin, Linux & OSX but **not** Solaris, but it could be added).

A complete `Makefile` is included for both `imsasim` and `z80sim`. From these it should be very easy to create the required `Makefile` for the other machines.

This has only been tested on OSX. So although I have made every effort to pick up the differences for each platform, I haven't been able to test that I got it right.

If you haven't previously seen the `*.d` files that are created during a build, these are common with other projects I am using. Each `*.d` file is a makefile for the given source, that gives all its dependencies as generated by `gcc -MM` (so maybe this won't be compatible with BSD (tba)). This means no more need to update the makefile because another header has been included.

With the above caveat regarding the addition of `-I.` to existing Makefiles, this build system can coexist with the existing build system unless further, expected changes are made that make the source files incompatible with the current system.

The included source file changes that require the addition of `-I.` to existing Makefiles is because occurrences of 
```
#include "sim.h"
#include "memory.h"
#include "config.h"
```
needed to be changed to 
```
#include <sim.h>
#include <memory.h>
#include <config.h>
```
This has avoided the need to move any source or headers into different directories. These 3 files need to be "overridden" by the files from the machine being compiled and the use of `#include <>` was the simplest way I could see how.

> Note: the previous use of  the system header file with `#include <memory.h>` in `z80sim/srcsim/simctl.c` and`z80sim/srcsim/simctl.c` I believe was in error and has been updated to `#include <string.h>` and covers the use of `memset()` in these modules. On the other hand it is not good that we have ended up with a "project" header file named `memory.h` as this could in future be a conflict with the system header file of the same name.

dmcnaugh commented:
Thanks for the review and feedback. 
Sorry that I went down a one-way, dead-end with the idea of using <inc.h> conventions. These have been reverted to "inc.h" in this PR.

Also I have changed the single makefile name to `GNUmakefile` to overcome the assumption that all platforms will have `GNU make` (raised in your comments in #53).

The solution with the least impact follows.

The main complication comes from having both the "core" files and the machine specific files in the same directory for the `z80sim` source. A further solution would be to seperate the "core" files into their own path eg. `z80pack/simcore` and have the z80sim machine build structured like all the other machines and symlink-ing the "core" files for the existing build system.

The solution I have included here is simpler. Only the 3 files that have to be "over-ridden" by each machine (`config.h`, `memory.h`, `sim.h`) have been moved into a new directory `z80sim/srcsim/local`. The `z80sim` existing makefiles need to have `-iquote .` added to the `gcc` build command and the paths to these three files make explicit `local/_filename_.h` (see `Makefile.osx` for an example). There is no such change required for the makefiles for other machines.

This way both build systems can co-exist.
- continue to use `make -f Makefile.os_platform` if preferred or `GNU make` isn't available
- using `make` alone will only work if `GNU make` is available which finds `GNUmakefile` by default

udo-munk commented:
No problem, we'll just get this better.

For someone not familiar with the logical layout of the components the z80sim/srcsim/local directory won't make much sense. Needs a while to figure that it is only to overcome limitations in build tools.

Of course it is much better to move the simulation core and then have the same layout for z80sim as with any other machine. Is that OK for you to modify the PR again before merging?

We can keep both build systems then for a while and if there aren't complains with the new one, remove the old one for good.

dmcnaugh commented:
OK. I've moved the "core" files to `z80core/`, reverted the creation of a `local` directory in `z80sim/simsrc` and updated `z80sim` to build like other machines (but only so far in `Makefile.osx`)

Changes to all the other `Makefile.os_platform` files for **z80sim** and all the other machines is very simple:
- make sure `-iquote .` ends up included in CFLAGS (probably by adding it to INCL if it exists)
- copy the `lnsrc` file from `z80sim/simsrc` to make sure all the symlinks to the "core" files now point to `../../z80core/*.[ch]`
